miner/miner.go

```diff
@@ -167,25 +176,32 @@ func (m *Miner) mine(ctx context.Context) {
 		}
 		lastBase = *base
 
-		b, err := m.mineOne(ctx, base)
-		if err != nil {
-			log.Errorf("mining block failed: %s", err)
-			log.Warn("waiting 400ms before attempting to mine a block")
-			time.Sleep(400 * time.Millisecond)
-			continue
+		blks := make([]*types.BlockMsg, 0)
+
+		for _, addr := range addrs {
+			b, err := m.mineOne(ctx, addr, base)
+			if err != nil {
+				log.Errorf("mining block failed: %s", err)
+				continue
+			}
+			if b != nil {
+				blks = append(blks, b)
+			}
 		}
 
-		if b != nil {
-			btime := time.Unix(int64(b.Header.Timestamp), 0)
+		if len(blks) != 0 {
+			btime := time.Unix(int64(blks[0].Header.Timestamp), 0)
 			if time.Now().Before(btime) {
 				time.Sleep(time.Until(btime))
 			} else {
 				log.Warnw("mined block in the past", "block-time", btime,
 					"time", time.Now(), "duration", time.Now().Sub(btime))
 			}
 
-			if err := m.api.SyncSubmitBlock(ctx, b); err != nil {
-				log.Errorf("failed to submit newly mined block: %s", err)
+			for _, b := range blks {
+				if err := m.api.SyncSubmitBlock(ctx, b); err != nil {
```

Comment:
This should make really, really sure that only one block per worker key is submitted